File: prradar/services/evaluation_service.py
"""Evaluation service - core rule evaluation logic using Claude Agent SDK.

This service handles the low-level details of evaluating code against rules
using the Claude Agent SDK with structured outputs.

Used by:
    - commands/agent/evaluate.py (batch evaluation command)
    - commands/agent/analyze.py (interactive pipeline)
"""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from claude_agent_sdk import ClaudeAgentOptions, ResultMessage, query

from prradar.domain.agent_outputs import RuleEvaluation
from prradar.domain.evaluation_task import EvaluationTask
from prradar.services.phase_sequencer import PhaseSequencer, PipelinePhase


# ============================================================
# Constants
# ============================================================

DEFAULT_MODEL = "claude-sonnet-4-20250514"

EVALUATION_PROMPT_TEMPLATE = """You are a code reviewer evaluating whether code violates a specific rule.

## Rule: {rule_name}

{rule_description}

### Rule Details

{rule_content}

## Focus Area: {focus_area_description}

File: {file_path}
Lines: {start_line}-{end_line}

**Important:** Only evaluate the code within the focus area boundaries shown below.
Ignore any surrounding code in the diff hunk.

**Code to review:**

```diff
{diff_content}
```

## Instructions

Analyze the code changes shown in the diff and determine if they violate the rule.

Focus ONLY on the added/changed lines (lines starting with `+`). Context lines (no prefix or starting with `-`) are provided for understanding but should not be evaluated for violations.

Consider:
1. Does the new or modified code violate the rule?
2. How severe is the violation (1-10 scale)?

For the comment field: If the rule includes a "GitHub Comment" section, use that exact text as your comment unless there is critical context-specific information that must be added. Keep comments concise.

Be precise about the file path and line number where any violation occurs.
"""


# ============================================================
# Domain Models
# ============================================================


@dataclass
class EvaluationResult:
    """Result of evaluating a single task."""

    task_id: str
    rule_name: str
    rule_file_path: str
    file_path: str
    evaluation: RuleEvaluation
    model_used: str
    duration_ms: int
    cost_usd: float | None

    # --------------------------------------------------------
    # Serialization
    # --------------------------------------------------------

    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "task_id": self.task_id,
            "rule_name": self.rule_name,
            "rule_file_path": self.rule_file_path,
            "file_path": self.file_path,
            "evaluation": self.evaluation.to_dict(),
            "model_used": self.model_used,
            "duration_ms": self.duration_ms,
            "cost_usd": self.cost_usd,
        }


# ============================================================
# Evaluation Logic
# ============================================================


async def evaluate_task(task: EvaluationTask) -> EvaluationResult:
    """Evaluate a single task using Claude Agent SDK.

    Args:
        task: The evaluation task containing rule and code segment

    Returns:
        EvaluationResult with the evaluation outcome
    """
    # Determine model to use (from rule or default)
    model = task.model or DEFAULT_MODEL

    # Build prompt from template using focused content
    focused_content = task.focus_area.get_focused_content()
    prompt = EVALUATION_PROMPT_TEMPLATE.format(
        rule_name=task.rule.name,
        rule_description=task.rule.description,
        rule_content=task.rule.content,
        focus_area_description=task.focus_area.description,
        file_path=task.focus_area.file_path,
        start_line=task.focus_area.start_line,
        end_line=task.focus_area.end_line,
        diff_content=focused_content,
    )

    # Configure structured output
    # Note: Don't use max_turns=1 as it may prevent the agent from completing
    # the structured output generation
    options = ClaudeAgentOptions(
        model=model,
        output_format={
            "type": "json_schema",
            "schema": RuleEvaluation.json_schema(),
        },
    )

    # Call Claude Agent SDK
    evaluation_data: dict | None = None
    duration_ms = 0
    cost_usd: float | None = None

    async for message in query(prompt=prompt, options=options):
        if isinstance(message, ResultMessage):
            # Capture structured output
            if message.structured_output:
                evaluation_data = message.structured_output
            duration_ms = message.duration_ms
            cost_usd = message.total_cost_usd

    # Check if we got valid structured output
    if evaluation_data:
        # Ensure file_path and line_number are populated from task if not in output
        if not evaluation_data.get("file_path"):
            evaluation_data["file_path"] = task.focus_area.file_path
        if not evaluation_data.get("line_number"):
            evaluation_data["line_number"] = task.focus_area.start_line

        evaluation = RuleEvaluation.from_dict(evaluation_data)
    else:
        evaluation = RuleEvaluation(
            violates_rule=False,
            score=1,
            comment="Evaluation failed - no structured output returned",
            file_path=task.focus_area.file_path,
            line_number=task.focus_area.start_line,
        )

    return EvaluationResult(
        task_id=task.task_id,
        rule_name=task.rule.name,
        rule_file_path=task.rule.file_path,
        file_path=task.focus_area.file_path,
        evaluation=evaluation,
        model_used=model,
        duration_ms=duration_ms,
        cost_usd=cost_usd,
    )


async def run_batch_evaluation(
    tasks: list[EvaluationTask],
    output_dir: Path,
    on_result: Callable[[int, int, EvaluationResult], None] | None = None,
) -> list[EvaluationResult]:
    """Run evaluations for all tasks.

    Core service method - handles evaluation and file I/O but no printing.
    Progress display is delegated to the caller via callback.

    Args:
        tasks: Tasks to evaluate
        output_dir: Where to save results (creates 'evaluations' subdirectory)
        on_result: Optional callback for progress reporting. Called with
                   (index, total, result) after each evaluation completes.
                   Index is 1-based. Service does NOT print progress - caller
                   handles display via this callback.

    Returns:
        List of all evaluation results
    """
    evaluations_dir = PhaseSequencer.ensure_phase_dir(output_dir, PipelinePhase.EVALUATIONS)

    results: list[EvaluationResult] = []
    total = len(tasks)

    for i, task in enumerate(tasks, 1):
        result = await evaluate_task(task)
        results.append(result)

        # Save evaluation result to file
        result_path = evaluations_dir / f"{task.task_id}.json"
        result_path.write_text(json.dumps(result.to_dict(), indent=2))

        # Notify caller of progress (UI layer handles printing)
        if on_result:
            on_result(i, total, result)

    return results

File: PRRadarLibrary/Sources/services/PRRadarCLIService/EvaluationService.swift
import Foundation
import PRRadarConfigService
import PRRadarModels

public struct EvaluationService: Sendable {
    private let bridgeClient: ClaudeBridgeClient

    private static let defaultModel = "claude-sonnet-4-20250514"

    private static let evaluationPromptTemplate = """
    You are a code reviewer evaluating whether code violates a specific rule.

    ## Rule: {rule_name}

    {rule_description}

    ### Rule Details

    {rule_content}

    ## Focus Area: {focus_area_description}

    File: {file_path}
    Lines: {start_line}-{end_line}

    **Important:** Only evaluate the code within the focus area boundaries shown below.
    Ignore any surrounding code in the diff hunk.

    **Code to review:**

    ```diff
    {diff_content}
    ```

    ## Codebase Context

    The PR branch is checked out locally at: {repo_path}
    You have full access to the codebase for additional context.

    - For rules that evaluate isolated patterns (naming conventions, signature
      format), the focus area content above is typically sufficient.
    - For rules that evaluate broader concerns (architecture, client usage,
      integration patterns), explore the codebase as needed. For example,
      search for callers of a method, check how similar patterns are used
      elsewhere, or read surrounding code for context.

    Use your judgment: explore when it would improve the quality of your
    review, but don't explore unnecessarily for simple pattern checks.

    ## Instructions

    Analyze the code changes shown in the diff and determine if they violate the rule.

    Focus ONLY on the added/changed lines (lines starting with `+`). Context lines \
    (no prefix or starting with `-`) are provided for understanding but should not be \
    evaluated for violations.

    Consider:
    1. Does the new or modified code violate the rule?
    2. How severe is the violation (1-10 scale)?

    For the comment field: If the rule includes a "GitHub Comment" section, use that \
    exact text as your comment unless there is critical context-specific information \
    that must be added. Keep comments concise.

    Be precise about the file path and line number where any violation occurs.
    """

    // Schema stored as nonisolated(unsafe) to avoid [String: Any] Sendable issues
    nonisolated(unsafe) private static let evaluationOutputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "violates_rule": [
                "type": "boolean",
                "description": "Whether the code violates the rule",
            ],
            "score": [
                "type": "integer",
                "minimum": 1,
                "maximum": 10,
                "description": "Severity score: 1-4 minor, 5-7 moderate, 8-10 severe",
            ],
            "comment": [
                "type": "string",
                "description": "The GitHub comment to post. If the rule includes a 'GitHub Comment' section, use that exact format unless there is critical context-specific information to add. Keep it concise.",
            ],
            "file_path": [
                "type": "string",
                "description": "Path to the file containing the code",
            ],
            "line_number": [
                "type": ["integer", "null"],
                "description": "Specific line number of the violation, if applicable",
            ],
        ] as [String: Any],
        "required": ["violates_rule", "score", "comment"],
    ]

    public init(bridgeClient: ClaudeBridgeClient) {
        self.bridgeClient = bridgeClient
    }

    /// Evaluate a single task using Claude via the bridge.
    public func evaluateTask(_ task: EvaluationTaskOutput, repoPath: String) async throws -> RuleEvaluationResult {
        let model = task.rule.model ?? Self.defaultModel
        let focusedContent = task.focusArea.getFocusedContent()

        let prompt = Self.evaluationPromptTemplate
            .replacingOccurrences(of: "{rule_name}", with: task.rule.name)
            .replacingOccurrences(of: "{rule_description}", with: task.rule.description)
            .replacingOccurrences(of: "{rule_content}", with: task.rule.content)
            .replacingOccurrences(of: "{focus_area_description}", with: task.focusArea.description)
            .replacingOccurrences(of: "{file_path}", with: task.focusArea.filePath)
            .replacingOccurrences(of: "{start_line}", with: String(task.focusArea.startLine))
            .replacingOccurrences(of: "{end_line}", with: String(task.focusArea.endLine))
            .replacingOccurrences(of: "{diff_content}", with: focusedContent)
            .replacingOccurrences(of: "{repo_path}", with: repoPath)

        let request = BridgeRequest(
            prompt: prompt,
            model: model,
            tools: ["Read", "Grep", "Glob"],
            cwd: repoPath,
            outputSchema: Self.evaluationOutputSchema
        )

        var bridgeResult: BridgeResult?
        for try await event in bridgeClient.stream(request) {
            switch event {
            case .text(let content):
                for textLine in content.components(separatedBy: "\n") {
                    print("      \(textLine)", terminator: "\n")
                }
            case .toolUse:
                break
            case .result(let result):
                bridgeResult = result
            }
        }

        guard let bridgeResult else {
            throw ClaudeBridgeError.noResult
        }

        let evaluation: RuleEvaluation
        if let dict = bridgeResult.outputAsDictionary() {
            let filePath = dict["file_path"] as? String ?? task.focusArea.filePath
            let lineNumber = dict["line_number"] as? Int ?? task.focusArea.startLine
            evaluation = RuleEvaluation(
                violatesRule: dict["violates_rule"] as? Bool ?? false,
                score: dict["score"] as? Int ?? 1,
                comment: dict["comment"] as? String ?? "Evaluation completed",
                filePath: filePath,
                lineNumber: lineNumber
            )
        } else {
            evaluation = RuleEvaluation(
                violatesRule: false,
                score: 1,
                comment: "Evaluation failed - no structured output returned",
                filePath: task.focusArea.filePath,
                lineNumber: task.focusArea.startLine
            )
        }

        return RuleEvaluationResult(
            taskId: task.taskId,
            ruleName: task.rule.name,
            ruleFilePath: "",
            filePath: task.focusArea.filePath,
            evaluation: evaluation,
            modelUsed: model,
            durationMs: bridgeResult.durationMs,
            costUsd: bridgeResult.costUsd
        )
    }

    /// Run evaluations for all tasks, writing results to the evaluations directory.
    public func runBatchEvaluation(
        tasks: [EvaluationTaskOutput],
        outputDir: String,
        repoPath: String,
        onStart: ((Int, Int, EvaluationTaskOutput) -> Void)? = nil,
        onResult: ((Int, Int, RuleEvaluationResult) -> Void)? = nil
    ) async throws -> [RuleEvaluationResult] {
        let evalsDir = "\(outputDir)/\(PRRadarPhase.evaluations.rawValue)"
        try FileManager.default.createDirectory(atPath: evalsDir, withIntermediateDirectories: true)

        var results: [RuleEvaluationResult] = []
        let total = tasks.count

        for (i, task) in tasks.enumerated() {
            let index = i + 1
            onStart?(index, total, task)

            let result = try await evaluateTask(task, repoPath: repoPath)
            results.append(result)

            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(result)
            let resultPath = "\(evalsDir)/\(DataPathsService.dataFilePrefix)\(task.taskId).json"
            try data.write(to: URL(fileURLWithPath: resultPath))

            onResult?(index, total, result)
        }

        return results
    }
}
